codec.go

```diff
@@ -31,6 +31,22 @@ type baseCodec interface {
 	Unmarshal(data []byte, v interface{}) error
 }
 
+// A reusableBaseCodec is exactly like a baseCodec, but also requires a
+// ReturnBuffer method to be implemented. Once a Marshal caller is done with
+// the returned byte buffer, they can choose to return it back to the encoding
+// library for re-use using this method.
+type reusableBaseCodec interface {
+	baseCodec
```

Comment:
This is unnecessary since you never use the baseCodec functions inside the type assertions checking for the `ReturnBuffer` method.  Delete and rename to `bufferReturner`?

Reply:
Done.